File: nautilus_trader/adapters/alpaca/factories.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2026 Andrew Crum. All rights reserved.
#  https://github.com/agcrum
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
"""
Alpaca adapter factory functions.

Provides factories for creating Alpaca data and execution clients that integrate
with NautilusTrader's event system.
"""

from __future__ import annotations

import asyncio
from functools import lru_cache

from nautilus_trader.adapters.alpaca.data import AlpacaDataClient
from nautilus_trader.adapters.alpaca.execution import AlpacaExecutionClient
from nautilus_trader.adapters.alpaca.providers import AlpacaInstrumentProvider
from nautilus_trader.cache.cache import Cache
from nautilus_trader.common.component import LiveClock
from nautilus_trader.common.component import MessageBus
from nautilus_trader.core import nautilus_pyo3
from nautilus_trader.live.factories import LiveDataClientFactory
from nautilus_trader.live.factories import LiveExecClientFactory


@lru_cache(1)
def get_cached_alpaca_http_client(
    api_key: str | None = None,
    api_secret: str | None = None,
    paper_trading: bool = True,
    base_url: str | None = None,
    timeout_secs: int | None = None,
    proxy_url: str | None = None,
) -> nautilus_pyo3.AlpacaHttpClient:
    """
    Cache and return an Alpaca HTTP client with the given credentials.

    If a cached client with matching parameters already exists, the cached client will be returned.

    Parameters
    ----------
    api_key : str, optional
        The API key for the client.
    api_secret : str, optional
        The API secret for the client.
    paper_trading : bool, default True
        If the client should use paper trading endpoints.
    base_url : str, optional
        The base URL for the API endpoints.
    timeout_secs : int, optional
        The timeout (seconds) for HTTP requests.
    proxy_url : str, optional
        The proxy URL for HTTP requests.

    Returns
    -------
    AlpacaHttpClient

    """
    # Determine environment
    environment = (
        nautilus_pyo3.AlpacaEnvironment.Paper
        if paper_trading
        else nautilus_pyo3.AlpacaEnvironment.Live
    )

    return nautilus_pyo3.AlpacaHttpClient(
        environment=environment,
        api_key=api_key or "",
        api_secret=api_secret or "",
        timeout_secs=timeout_secs,
        proxy_url=proxy_url,
    )


@lru_cache(1)
def get_cached_alpaca_instrument_provider(
    client: nautilus_pyo3.AlpacaHttpClient,
    config=None,  # InstrumentProviderConfig
) -> AlpacaInstrumentProvider:
    """
    Cache and return an Alpaca instrument provider.

    If a cached provider already exists, then that provider will be returned.

    Parameters
    ----------
    client : AlpacaHttpClient
        The Alpaca HTTP client.
    config : InstrumentProviderConfig, optional
        The instrument provider configuration, by default None.

    Returns
    -------
    AlpacaInstrumentProvider

    """
    return AlpacaInstrumentProvider(
        client=client,
        config=config,
    )


class AlpacaLiveDataClientFactory(LiveDataClientFactory):
    """
    Provides an Alpaca live data client factory.
    """

    @staticmethod
    def create(  # type: ignore
        loop: asyncio.AbstractEventLoop,
        name: str,
        config,  # AlpacaDataClientConfig
        msgbus: MessageBus,
        cache: Cache,
        clock: LiveClock,
    ) -> AlpacaDataClient:
        """
        Create a new Alpaca data client.

        Parameters
        ----------
        loop : asyncio.AbstractEventLoop
            The event loop for the client.
        name : str
            The custom client ID.
        config : AlpacaDataClientConfig
            The client configuration.
        msgbus : MessageBus
            The message bus for the client.
        cache : Cache
            The cache for the client.
        clock : LiveClock
            The clock for the client.

        Returns
        -------
        AlpacaDataClient

        """
        client: nautilus_pyo3.AlpacaHttpClient = get_cached_alpaca_http_client(
            api_key=config.api_key,
            api_secret=config.api_secret,
            paper_trading=config.paper_trading,
            base_url=config.base_url_http,
            timeout_secs=config.http_timeout_secs,
            proxy_url=config.proxy_url,
        )

        provider = get_cached_alpaca_instrument_provider(
            client=client,
            config=config.instrument_provider,
        )

        return AlpacaDataClient(
            loop=loop,
            client=client,
            msgbus=msgbus,
            cache=cache,
            clock=clock,
            instrument_provider=provider,
            config=config,
            name=name,
        )


class AlpacaLiveExecClientFactory(LiveExecClientFactory):
    """
    Provides an Alpaca live execution client factory.
    """

    @staticmethod
    def create(  # type: ignore
        loop: asyncio.AbstractEventLoop,
        name: str,
        config,  # AlpacaExecClientConfig
        msgbus: MessageBus,
        cache: Cache,
        clock: LiveClock,
    ) -> AlpacaExecutionClient:
        """
        Create a new Alpaca execution client.

        Parameters
        ----------
        loop : asyncio.AbstractEventLoop
            The event loop for the client.
        name : str
            The custom client ID.
        config : AlpacaExecClientConfig
            The client configuration.
        msgbus : MessageBus
            The message bus for the client.
        cache : Cache
            The cache for the client.
        clock : LiveClock
            The clock for the client.

        Returns
        -------
        AlpacaExecutionClient

        """
        client: nautilus_pyo3.AlpacaHttpClient = get_cached_alpaca_http_client(
            api_key=config.api_key,
            api_secret=config.api_secret,
            paper_trading=config.paper_trading,
            base_url=config.base_url_http,
            timeout_secs=config.http_timeout_secs,
            proxy_url=config.proxy_url,
        )

        provider = get_cached_alpaca_instrument_provider(
            client=client,
            config=None,
        )

        return AlpacaExecutionClient(
            loop=loop,
            client=client,
            msgbus=msgbus,
            cache=cache,
            clock=clock,
            instrument_provider=provider,
            config=config,
            name=name,
        )

File: crates/adapters/alpaca/src/factories.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2026 Andrew Crum. All rights reserved.
//  https://github.com/agcrum
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Factory functions for creating Alpaca clients and providers.

use std::{env, sync::Arc};

use crate::{
    common::enums::AlpacaEnvironment,
    config::{AlpacaDataClientConfig, AlpacaExecClientConfig, AlpacaInstrumentProviderConfig},
    error::{AlpacaError, Result},
    http::client::AlpacaHttpClient,
    providers::AlpacaInstrumentProvider,
};

/// Factory for creating Alpaca HTTP clients with optional credential loading from environment.
#[derive(Debug)]
pub struct AlpacaHttpClientFactory;

impl AlpacaHttpClientFactory {
    /// Creates a new HTTP client with the given configuration.
    ///
    /// If API credentials are not provided in the config, attempts to read them from
    /// environment variables `ALPACA_API_KEY` and `ALPACA_API_SECRET`.
    ///
    /// # Errors
    ///
    /// Returns an error if credentials are not provided and not found in environment.
    pub fn create(config: &AlpacaDataClientConfig) -> Result<AlpacaHttpClient> {
        let api_key = config
            .api_key
            .clone()
            .or_else(|| env::var("ALPACA_API_KEY").ok())
            .ok_or_else(|| {
                AlpacaError::InvalidRequest(
                    "API key not provided and ALPACA_API_KEY environment variable not set"
                        .to_string(),
                )
            })?;

        let api_secret = config
            .api_secret
            .clone()
            .or_else(|| env::var("ALPACA_API_SECRET").ok())
            .ok_or_else(|| {
                AlpacaError::InvalidRequest(
                    "API secret not provided and ALPACA_API_SECRET environment variable not set"
                        .to_string(),
                )
            })?;

        let environment = if config.paper_trading {
            AlpacaEnvironment::Paper
        } else {
            AlpacaEnvironment::Live
        };

        AlpacaHttpClient::new(
            environment,
            api_key,
            api_secret,
            config.http_timeout_secs,
            config.proxy_url.clone(),
        )
        .map_err(|e| AlpacaError::InvalidRequest(format!("Failed to create HTTP client: {:?}", e)))
    }

    /// Creates a new HTTP client for execution with the given configuration.
    ///
    /// # Errors
    ///
    /// Returns an error if credentials are not provided and not found in environment.
    pub fn create_for_exec(config: &AlpacaExecClientConfig) -> Result<AlpacaHttpClient> {
        let api_key = config
            .api_key
            .clone()
            .or_else(|| env::var("ALPACA_API_KEY").ok())
            .ok_or_else(|| {
                AlpacaError::InvalidRequest(
                    "API key not provided and ALPACA_API_KEY environment variable not set"
                        .to_string(),
                )
            })?;

        let api_secret = config
            .api_secret
            .clone()
            .or_else(|| env::var("ALPACA_API_SECRET").ok())
            .ok_or_else(|| {
                AlpacaError::InvalidRequest(
                    "API secret not provided and ALPACA_API_SECRET environment variable not set"
                        .to_string(),
                )
            })?;

        let environment = if config.paper_trading {
            AlpacaEnvironment::Paper
        } else {
            AlpacaEnvironment::Live
        };

        AlpacaHttpClient::new(
            environment,
            api_key,
            api_secret,
            config.http_timeout_secs,
            config.proxy_url.clone(),
        )
        .map_err(|e| AlpacaError::InvalidRequest(format!("Failed to create HTTP client: {:?}", e)))
    }
}

/// Factory for creating Alpaca instrument providers.
#[derive(Debug)]
pub struct AlpacaInstrumentProviderFactory;

impl AlpacaInstrumentProviderFactory {
    /// Creates a new instrument provider with the given HTTP client and configuration.
    ///
    /// # Errors
    ///
    /// Returns an error if the configuration is invalid.
    pub fn create(
        client: AlpacaHttpClient,
        config: Option<AlpacaInstrumentProviderConfig>,
    ) -> Result<AlpacaInstrumentProvider> {
        let provider_config = if let Some(cfg) = config {
            crate::providers::AlpacaInstrumentProviderConfig {
                asset_classes: cfg.asset_classes.unwrap_or_default(),
                option_underlying_symbols: cfg.option_underlying_symbols,
                venue: nautilus_model::identifiers::Venue::from("ALPACA"),
                log_warnings: cfg.log_warnings,
            }
        } else {
            crate::providers::AlpacaInstrumentProviderConfig::default()
        };

        Ok(AlpacaInstrumentProvider::new(client, provider_config))
    }
}

/// Cached HTTP client instance for reuse across data and execution clients.
static mut CACHED_HTTP_CLIENT: Option<Arc<AlpacaHttpClient>> = None;
static mut CACHED_CLIENT_CONFIG_HASH: u64 = 0;

/// Returns a cached HTTP client, creating one if it doesn't exist or if the config has changed.
///
/// This function uses a simple caching mechanism to avoid creating multiple HTTP clients
/// with the same configuration. Note: This is not thread-safe and should only be used
/// in single-threaded contexts or with external synchronization.
///
/// # Errors
///
/// Returns an error if the client cannot be created.
pub fn get_cached_http_client(config: &AlpacaDataClientConfig) -> Result<Arc<AlpacaHttpClient>> {
    unsafe {
        // Simple hash of config for cache invalidation
        let config_hash = compute_config_hash(config);

        let cached_hash = std::ptr::read(std::ptr::addr_of!(CACHED_CLIENT_CONFIG_HASH));
        let client_ptr = std::ptr::addr_of!(CACHED_HTTP_CLIENT);

        if cached_hash != config_hash || std::ptr::read(client_ptr).is_none() {
            let client = AlpacaHttpClientFactory::create(config)?;
            std::ptr::write(std::ptr::addr_of_mut!(CACHED_HTTP_CLIENT), Some(Arc::new(client)));
            std::ptr::write(std::ptr::addr_of_mut!(CACHED_CLIENT_CONFIG_HASH), config_hash);
        }

        Ok(std::ptr::read(client_ptr)
            .expect("Client should be initialized"))
    }
}

/// Returns a cached HTTP client for execution, creating one if needed.
///
/// # Errors
///
/// Returns an error if the client cannot be created.
pub fn get_cached_http_client_for_exec(
    config: &AlpacaExecClientConfig,
) -> Result<Arc<AlpacaHttpClient>> {
    unsafe {
        // For execution clients, we use the same cache mechanism
        // Convert exec config to data config for hashing
        let data_config = AlpacaDataClientConfig {
            api_key: config.api_key.clone(),
            api_secret: config.api_secret.clone(),
            paper_trading: config.paper_trading,
            base_url_http: config.base_url_http.clone(),
            http_timeout_secs: config.http_timeout_secs,
            ..Default::default()
        };

        let config_hash = compute_config_hash(&data_config);

        let cached_hash = std::ptr::read(std::ptr::addr_of!(CACHED_CLIENT_CONFIG_HASH));
        let client_ptr = std::ptr::addr_of!(CACHED_HTTP_CLIENT);

        if cached_hash != config_hash || std::ptr::read(client_ptr).is_none() {
            let client = AlpacaHttpClientFactory::create_for_exec(config)?;
            std::ptr::write(std::ptr::addr_of_mut!(CACHED_HTTP_CLIENT), Some(Arc::new(client)));
            std::ptr::write(std::ptr::addr_of_mut!(CACHED_CLIENT_CONFIG_HASH), config_hash);
        }

        Ok(std::ptr::read(client_ptr)
            .expect("Client should be initialized"))
    }
}

/// Computes a simple hash of the config for cache invalidation.
fn compute_config_hash(config: &AlpacaDataClientConfig) -> u64 {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    let mut hasher = DefaultHasher::new();
    config.api_key.hash(&mut hasher);
    config.api_secret.hash(&mut hasher);
    config.paper_trading.hash(&mut hasher);
    config.base_url_http.hash(&mut hasher);
    hasher.finish()
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;


    #[rstest]
    fn test_config_hash_same_for_identical_configs() {
        let config1 = AlpacaDataClientConfig::new()
            .with_credentials("key".to_string(), "secret".to_string())
            .with_paper_trading(true);

        let config2 = AlpacaDataClientConfig::new()
            .with_credentials("key".to_string(), "secret".to_string())
            .with_paper_trading(true);

        assert_eq!(compute_config_hash(&config1), compute_config_hash(&config2));
    }

    #[rstest]
    fn test_config_hash_different_for_different_configs() {
        let config1 = AlpacaDataClientConfig::new()
            .with_credentials("key1".to_string(), "secret1".to_string())
            .with_paper_trading(true);

        let config2 = AlpacaDataClientConfig::new()
            .with_credentials("key2".to_string(), "secret2".to_string())
            .with_paper_trading(true);

        assert_ne!(compute_config_hash(&config1), compute_config_hash(&config2));
    }

    #[rstest]
    fn test_instrument_provider_factory() {
        // Note: This test would require a valid HTTP client
        // In a real scenario, you'd use a mock client or integration test
        let client = AlpacaHttpClient::new(
            AlpacaEnvironment::Paper,
            "test_key".to_string(),
            "test_secret".to_string(),
            Some(30),
            None,
        )
        .expect("Failed to create client");

        let provider_config = AlpacaInstrumentProviderConfig::default();
        let result = AlpacaInstrumentProviderFactory::create(client, Some(provider_config));

        assert!(result.is_ok());
    }
}
